hal/armv7m/mpu.c

```diff
@@ -0,0 +1,322 @@
+
+	mpu_common.hal.mpu.region[idx].rasr =
+		((((attr & mAttrExec) == 0) & 1ul) << 28) |
+		((ap & 0x7ul) << 24) |
+		((tex & 0x7ul) << 19) |
+		((((attr & mAttrShareable) != 0) & 1ul) << 18) |
+		((((attr & mAttrCacheable) != 0) & 1ul) << 17) |
+		((((attr & mAttrBufferable) != 0) & 1ul) << 16) |
+		((srdMask & 0xfful) << 8) |
+		(((sizeBit - 1) & 0x1ful) << 1) |
+		(enable != 0);
+
+	return EOK;
+}
+
+
+/* Find best alignment for map and setup a region */
+static int mpu_regionBestFit(u8 region, u32 addr, u32 size, u32 attr, u8 enable, u32 *allocSize)
+{
+	u32 srBase, srSize, alignMask;
+
+	u8 alignBits = _ctz(addr);
+	u8 srdMask = 0, bit = _ctz(size);
+
+	if (bit < alignBits)
+		alignBits = bit;
+
+	alignMask = alignBits < 32 ? ~((1u << alignBits) - 1) : 0;
+
+	if (alignBits < 5) {
+		*allocSize = 1u << alignBits;
+	}
+	else {
+		alignBits += 3;
+
+		if (alignBits >= 32) {
+			alignBits = 32;
+			alignMask = 0;
+		}
+		else {
+			alignMask = ~((1u << alignBits) - 1);
+		}
+
+		*allocSize = 0;
+		srBase = addr & alignMask;
+		srSize = 1u << (alignBits - 3);
+
+		for (bit = 0; bit < 8; bit++) {
+			if (srBase < addr || srBase + srSize > addr + size || *allocSize + srSize > size)
+				srdMask |= 1u << bit;
+			else
+				*allocSize += srSize;
+
+			srBase += srSize;
+		}
+	}
+
+	return mpu_regionSet(region, addr & alignMask, alignBits, attr, enable, srdMask);
+}
+
+
+/* Allocate MPU sub/regions for a map */
+static int mpu_regionAlloc(u8 *regCnt, u32 mapId, u32 addr, u32 size, u32 attr, u8 enable)
+{
+	u32 allocSize = 0;
+	u8 regCur = *regCnt;
+
+	while (size > 0 && (regCur - *regCnt) < 2) {
+		if (mpu_regionBestFit(regCur, addr, size, attr, enable, &allocSize) < 0)
+			return -EPERM;
+
+		if (allocSize > size)
+			return -EPERM;
+
+		addr += allocSize;
+		size -= allocSize;
+		regCur++;
+	}
+
+	if (size != 0)
+		return -EPERM;
+
+	for (;;) {
+		mpu_common.hal.mpu.map[*regCnt] = mapId;
+		if (*regCnt >= regCur)
+			break;
+		(*regCnt)++;
+	}
+
+	return EOK;
+}
+
+
+/* Allocate MPU sub/regions for a map */
+int mpu_mapAlloc(addr_t start, addr_t end, u32 attr, u32 id)
+{
+	int res;
+	if ((res = mpu_regionAlloc(&mpu_common.regCnt, id, start, end - start, attr, 1)) < 0) {
+		mpu_common.flags |= mpu_fError;
+		return res;
+	}
+
+	mpu_common.mapsCnt++;
+
+	return EOK;
+}
+
+
+/* Prints allocated MPU regions */
+void mpu_mapInfo(void)
+{
+	unsigned int i;
+	mpu_region_t *region;
+	char *name;
+
+	lib_printf(CONSOLE_BOLD "\n%-10s %-6s %-9s %-14s %-14s", "NAME", "REGION", "SUBREGION", "START", "END");
+	lib_printf("EN  XN  ACCESS P/U  TEX  S  C  B\n" CONSOLE_NORMAL);
+
+	for (i = 0; i < mpu_common.regCnt; i++) {
+		region = &mpu_common.hal.mpu.region[i];
+		if ((name = syspage_getMapName(mpu_common.hal.mpu.map[i])) == NULL)
+			name = "<invalid>";
+
+		mpu_regionPrint(name, region->rbar, region->rasr);
+	}
+
+	lib_printf("\nConfigured %d of %d MPU regions based on %d map definitions.\n",
+		mpu_common.regCnt, mpu_common.regMax, mpu_common.mapsCnt);
+}
+
```

Comment:
This function is redundant. All information about maps should be displayed by map command belonging to hardware independent part.